File: hash_set.py

#Implementing primary array and a secondary array at each index of primary array
#the values of secondary array will be boolean- gives O(1) search,remove,add


class MyHashSet:

    def __init__(self):
        #initialize a array of size 1000
        
        self.array=[None for i in range(1000)]
        self.buckets=1000
        self.bucketitems=1000
        

    def add(self, key: int) -> None:
        '''O(1) Time complexity
           O(1) Space complexity'''
        #first hash
        key1=key%self.buckets
        #second hash
        key2=key//self.bucketitems
        #check if element at key1 is None(no secondary array)
        if self.array[key1]==None:
            #if primary array index is one(edge case for valye=10^6)
            if key1==0:
                self.array[key1]=[False for i in range(1001)]
            else:
                self.array[key1]=[False for i in range(1000)]
        #set the value to boolean 
        self.array[key1][key2]=True
        

    def remove(self, key: int) -> None:
        '''O(1) Time complexity
           O(1) Space complexity'''
        key1=key%self.buckets
        key2=key//self.bucketitems
        if self.array[key1]:
            if self.array[key1][key2]==True:
                self.array[key1][key2]=False
            
        
        

    def contains(self, key: int) -> bool:
        '''O(1) Time complexity
           O(1) Space complexity'''
        key1=key%self.buckets
        key2=key//self.bucketitems
        if self.array[key1]==None:
            return False
        
        return self.array[key1][key2]
        
        


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)

File: src/main/java/HashSet.java
class MyHashSet {

    boolean[][] storage;
    int buckets;
    int bucketItems;
    /** Bucket and bucketItems is used for hashing algorithm.
     * Storage is a 2 d boolean array which stores true if value is present false if not.
     * Implemented this using double hashing*/
    public MyHashSet() {
        this.storage=new boolean[1000][];
        this.bucketItems=1000;
        this.buckets=1000;
    }

    public int bucket(int key){
        return key%buckets;
    }

    public int bucketItems(int key){
        return key/bucketItems;
    }

    public void add(int key) {
        int bucket=bucket(key);
        int bucketItem=bucketItems(key);
        if(storage[bucket]==null){
            if(bucket==0){
                storage[bucket]=new boolean[bucketItems+1];
            }else{
                storage[bucket]=new boolean[bucketItems];
            }
        }
        storage[bucket][bucketItem]=true;
    }

    public void remove(int key) {
        int bucket=bucket(key);
        int bucketItem=bucketItems(key);
        if(storage[bucket]!=null){
            storage[bucket][bucketItem]=false;
        }
    }

    /** Returns true if this set contains the specified element */
    public boolean contains(int key) {
        int bucket=bucket(key);
        int bucketItem=bucketItems(key);
        if(storage[bucket]!=null){
            if(storage[bucket][bucketItem]==true){
                return true;
            }
        }
        return false;

    }
}

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet obj = new MyHashSet();
 * obj.add(key);
 * obj.remove(key);
 * boolean param_3 = obj.contains(key);
 */